### PR TITLE
Add support for autotune limits in service_shard_update

### PR DESCRIPTION
**Pull request by karmeleon**
Added to PaaSTA at large in #3744. Also pulls in commit from #3829 to make min/max instance values optional and adds a new `--dry-run` param that skips committing and pushing changes to yelpsoa-configs.

Confirmed that the new kwargs do as expected, generating valid configs. If anyone knows a way to do this that doesn't involve a mess of nested if statements, I'm all ears lol

### CODE REVIEW

File: paasta_tools/contrib/service_shard_update.py
Comment (karmeleon) [inline paasta_tools/contrib/service_shard_update.py:200]: @nemacysts (from #3829) how would you suggest testing paasta's behavior with no max/min instances?

Comment (nemacysts) [reply paasta_tools/contrib/service_shard_update.py:200]: @karmeleon sorry, yesterday was a busy day!

i don't think we've setup the local paasta playground for autoscaling, but we should be able to create an instance in infrastage (or any test cluster) and see what happens

lemme do that real fast, actually :)

Comment (nemacysts) [reply paasta_tools/contrib/service_shard_update.py:200]: whoops - replied in the wrong place: https://github.com/Yelp/paasta/pull/3829#discussion_r1569091874

Comment (nemacysts) [reply paasta_tools/contrib/service_shard_update.py:200]: PAASTA-18283

Comment (karmeleon) [reply paasta_tools/contrib/service_shard_update.py:200]: thanks for checking! to clarify, we're totally okay with min/max instances both being at 1 for brand new shards, since they typically see traffic ramp up gradually enough for autotune to deal with it. do you think it's safe to ship this branch now, or should we wait for that ticket to be completed?

Comment (nemacysts) [reply paasta_tools/contrib/service_shard_update.py:200]: i'm trying to verify that autotune will do the "right" thing and setup min/max instances - but when i ran the rightsizer scripts locally it didn't seem to...that said, i'm not sure if it's cause i need to wait for splunk to have more data

Comment (nemacysts) [reply paasta_tools/contrib/service_shard_update.py:200]: whoops, sorry - forgot about this: i think i hallucinated that autotune would fixup the autoscaling config: my test instance has not had autotune add min/max instances correctly :(

Comment (nemacysts) [reply paasta_tools/contrib/service_shard_update.py:200]: (i.e., we'll either want to revert the {min,max}_instances changes here or tackle PAASTA-18283)

File: paasta_tools/contrib/service_shard_update.py
Comment (nemacysts) [inline paasta_tools/contrib/service_shard_update.py:48]: ++

(i'm also a fan of `--for-real` flags so that dry-run is the default, but that would be a little more annoying to rollout here)

File: paasta_tools/contrib/service_shard_update.py
Comment (nemacysts) [inline paasta_tools/contrib/service_shard_update.py:287]:
```suggestion
                        instance_config["autotune_limits"] = {}
```
is the key we want instead

---

might also be nice to do something like:
```python
limit_config = instance_config["autotune_limits"]
...
limit_config["cpus"]["min"] = args.autotune_min_cpus
...
```
so that we don't need to repeat `instance_config["autotune_limits"]` so many times :)

Comment (nemacysts) [reply paasta_tools/contrib/service_shard_update.py:287]: i think maybe we could trim this code down by doing something like:
```
limit_config["cpus"] = {
    "min": args.autotune_min_cpus,
    "max": args.autotune_max_cpus
}
limit_config["mem"] = {
    "min": args.autotune_min_mem,
    "max": args.autotune_max_mem
}
limit_config["disk"] = {
    "min": args.autotune_min_disk,
    "max": args.autotune_max_disk
}```
and then adding a processing step that removes any min/max keys with `None` values - and any resource (e.g., cpus, mem, disk) keys with no subkeys

(that said, maybe the code to do that processing step wouldn't look too different and doing this would be a waste of time :p)

**Review by nemacysts — COMMENTED** [review]
generally lgtm - we'll probably want to back out the min/max instance changes + fix the autotune limits typo, but otherwise i don't see any concerns here

**Comment by karmeleon** [comment]
@nemacysts sorry for the 6 months of suspense! kinda forgot about this PR until we ran into some super anemic ssr shards with 0.1 cpus causing hella timeouts on prod lol

**Review by nemacysts — APPROVED** [review]
lgtm - unless you'd like to add tests here, i'm assuming that you ran this locally and got the expected output :)

(and in any case, we'd probably wanna wait until monday to release this anyway :p)

File: paasta_tools/contrib/service_shard_update.py
Comment (nemacysts) [inline paasta_tools/contrib/service_shard_update.py:297]: (these list()'s aren't necessary, but i'm assuming they're were helpful in some way while you were debugging and converting from a view to a list shouldn't really impact perf in a meaningful way for this script)

Comment (karmeleon) [reply paasta_tools/contrib/service_shard_update.py:297]: unfortunately they are necessary under py3 :(

```
Traceback (most recent call last):
  File "paasta_tools/contrib/service_shard_update.py", line 350, in <module>
    main(args)
  File "paasta_tools/contrib/service_shard_update.py", line 297, in main
    for key in limit_config[resource]:
RuntimeError: dictionary changed size during iteration
```

**Comment by karmeleon** [comment]
yep, got the expected output :) will merge this next week

File: paasta_tools/contrib/service_shard_update.py
Comment (nemacysts) [reply paasta_tools/contrib/service_shard_update.py:297]: oh d'oh - i'm dumb: we're mutating the dict 

i blame friday-brain :p

**Pull request merged**
